posts/fantastic-simprocs.md

```diff
@@ -0,0 +1,198 @@
+- Check that `∣` is the usual natural number division.
+- Check that `a` and `b` are numerals.
+- Compute `b % a`.
+- If this quantity is zero, then return `True` together with the proof `Nat.dvd_eq_true_of_mod_eq_zero a b rfl`.
+- If this quantity isn't zero, then return `False` together with the proof `Nat.dvd_eq_false_of_mod_ne_zero a b rfl`.
+
+## Avoiding combinatorial explosion of lemmas: The `existsAndEq` simproc
+
+The `existsAndEq` simproc is designed to simplify expressions of the form `∃ x, ... ∧ x = a ∧ ...` where `a` is some quantity independent of `x` by removing the existential quantifier and replacing all occurences of `x` by `a`.
+
+```lean
+example (p : Nat → Prop) : ∃ x : Nat, p x ∧ x = 5 := by
+  simp only [existsAndEq]
+  -- Remaining goal: `⊢ p 5`
+
+example (p q : Nat → Prop) : ∃ x : Nat, p x ∧ x = 5 ∧ q x := by
+  simp only [existsAndEq]
+  -- Remaining goal: `⊢ p 5 ∧ q 5`
+```
+
+Intuitively, one can think of `existsAndEq` as a "parametric" lemma whose right hand side is allowed to vary shape depending on the left hand side.
+
+Roughly speaking, the way this metaprogram operates is a follows: whenever an expression of the form `∃ a, P a` with `P a = Q ∧ R`  is encountered:
+- Recursively traverse the expression of the predicate inside the existential quantifier to try and detect an equality `a = a'` by splitting any `∧` that is found along the way into its components.
+- If an equality is found, construct a proof that `P a` implies `a = a'`.
+- Construct a proof that `(∃ a, P a) = P a'` using the following theorem
+```lean
+theorem exists_of_imp_eq {α : Sort u} {p : α → Prop} (a : α) (h : ∀ b, p b → a = b) :
+    (∃ b, p b) = p a
+```
+
+## Performance optimisation: The `reduceIte` simproc
+
+The `reduceIte` simproc is designed to take expressions of the form `if P then a else b` (aka `ite P a b`) and replace them with `a` or `b`, depending on whether `P` simplify to `True` or `False`.
+
+```lean
+example : (if 37 * 0 then 1 else 2) = 1 := by
+  -- Works since `simp` can simplify `37 * 0` to `True`
+  -- because it knows the lemma `mul_zero a : a * 0 = 0`.
+  simp only [↓reduceIte]
+
+example (X : Type) (P : Prop) (a b : X) : (if P ∨ ¬ P then a else b) = a := by
+  -- Works since `simp` can simplify `P ∨ ¬ P` to `True`.
+  simp only [↓reduceIte]
+
+open Classical
+
+example : (if FermatLastTheorem then 1 else 2) = 1 := by
+  --This fails because `simp` can't simplify `FermatLastTheorem` to `True` or `False`
+  simp only [↓reduceIte]
+  -- Remaining goal: `⊢ (if FermatLastTheorem then 1 else 2) = 1`
+  sorry -- See https://imperialcollegelondon.github.io/FLT for how to solve this `sorry` ;)
+```
+
+The down arrow `↓` in `simp only [↓reduceIte]` indicates that `reduceIte` is being called as a *preprocedure*.
+This means that `reduceIte` is run *before* the expressions `a` and `b` are simplified.
```

Comment:
You could mention here that this is an example of why "simplification is performed from the inside-out" was a simplification (no pun intended), which a) gives a tiny bit more context on what it actually happening and b) makes it clearer that you're not self-contradicting